openlibrary/catalog/add_book/__init__.py

```diff
@@ -678,10 +686,340 @@ def load_data(rec, account_key=None):
+# so we'd compare `key`
+THING_EDITION_FIELDS: Final = {
+    'languages': 'key',
+    'links': 'url',
+    'translated_from': 'key',
+}
+
+
+def add_missing_things_and_thing_values_to_edition(
+    source_rec: dict,
+    edition: Edition,
+    thing_edition_fields: dict[str, str],
+    need_edition_save: bool,
+    overwrite: bool = False,
+) -> tuple[Edition, bool]:
+    """
+    Add missing lists and list values from source_rec to the target edition.
+
+    :param source_rec: A `rec` dictionary *returned from build_query()*.
+    :param edition: A the matched target edition that will be updated with any new values.
+    :param edition_fields: A list of field names, themselves containing lists, to copy.
+    :param need_edition_save: True if the edition has been modified and needs saving.
+    :param overwrite: If true, use source_rec values to overwrite those in edition.
+    :return: A tuple containing the (possibly modified) edition and a bool indicating
+        whether the edition needs to be saved.
+    """
+    for field, thing_key in thing_edition_fields.items():
+        if not source_rec.get(field):
+            continue
+
+        if overwrite:
+            edition[field] = source_rec[field]
+            need_edition_save = True
+            continue
+
+        # rec_edition has a JSON-like representation of some values that are Thing objects in the edition.
+        # Get the rec_edition values from the current field, and then add them if they're not in the edition.
+        if field in edition:
+            values_in_edition = [thing.get(thing_key) for thing in edition[field]]
+            values_to_append = [
+                item
+                for item in source_rec[field]
+                if item.get(thing_key) not in values_in_edition
+            ]
+
+            if values_to_append:
+                edition[field] += values_to_append
+                need_edition_save = True
+        else:
+            edition[field] = source_rec[field]
+            need_edition_save = True
+
+    return (edition, need_edition_save)
+
+
+def check_if_edition_should_be_overwritten(
+    edition: Edition, from_marc_record: bool
+) -> bool:
+    """
+    Certain editions should be overwritten with rec data because of their
+    pervasive low quality, such as v1 Promise Items.
+
+    :edition The matched Edition to evaluate for overwriting.
+    :from_marc_record True if the rec associated with this matched Edition
+        has a MARC record.
+    :returns True if the edition data should be overwritten.
+    """
+    # Trust MARC record data implicitly over v1 Promise Items.
+    try:
+        ia_promise_item = any(
+            record
+            for record in edition.get('source_records')
+            if record.split(":")[0] == "promise"
+        )
+        if ia_promise_item and edition.get('revision') == 1 and from_marc_record:
+            return True
+    except TypeError:
```

Comment:
What is this `TypeError` catching?

Also, this could return `False` early by checking `not from_marc_record`, or  a revision greater than 1, without having to iterate on source_records.   For a v=1 item there will only be one source record anyway, no point iterating over this if we don't have to.

Reply:
Oh, I really like this approach. Thanks! I went ahead and updated the PR to reflect this suggestion.

I took out the `try`/`except`, but it was for when `edition.get('source_records')` returned `None` and the list comprehension tried to iterate over `None`. Granted, I could have solved that with `edition.get('source_records', [])`. But it's all moot now. :D